### ISSUE TITLE
Can't find words in Kobo dictionary generated by dictgen

### ISSUE THREAD

**Issue by Ceiyne** [pgaskin/dictutil#16]
This is my first time using dictgen, so I apologize if this is actually user error.

I have been trying to convert JMdict to a Kobo dictionary.  I used Pyglossary to generate a df file from JMdict, and then used dictgen to create the Kobo dictionary.  After installing it (using the custom-dict folder), the Kobo saw the dictionary, but none of the words I tried to look up were able to be found.  These were all words that are present in the JMdict source file.

When I ran dictgen, I used the default options.  It ran without errors and said that it successfully wrote 190,800 entries.

I did a little troubleshooting but didn't come up with anything solid.  Here are a few notes from what I checked:

1. I looked at the df file that Pyglossary generated and it appeared to be in the correct format based on what I see on your documentation page. I also verified the entries I was trying to find in the book, and they were present in this file as well.
2. I looked at the zip file that dictgen generated and on the surface it looked like my other Kobo dictionaries.  It contained many files with filenames with two-character names like xy.html and those files contained unreadable data.
3. I looked at your existing issues but didn't see anything similar.  I saw one issue where your notes mentioned a "no words found" bug triggered by spaces in the dictionary filename, but I did not have spaces.  I tried a few different names to make sure it wasn't a naming issue, things like: dicthtml-test.zip and dicthtml-test-test.zip

Pyglossary is capable of creating Kobo dictionaries as well, so in case you were wondering why I didn't just do that... I tried that method but had issues there as well.  With Pyglossary the generated dictionary worked to some extent -- the Kobo would return the correct dictionary entries for many words.  But there were also a lot of words that could not be found despite being present in JMdict.  So, I thought I'd try working with dictgen instead.

**Comment by pgaskin** [comment]
If this is the same Japanese issue reported on PyGlossary, note that the Kobo dictionary implementation in PyGlossary was derived from dictutil 🙂.

If it's not, can you provide a few examples of words which aren't found?

**Comment by Ceiyne** [comment]
Yeah, it's basically the same one.  The one over there was from when I used Pyglossary to do the whole conversion, and the one here was when I used Pyglossary to make the df file and dictgen to create the dicthtml.zip.  

I'm on the latest (as far as I know) firmware, 4.25.15875.

**Comment by pgaskin** [comment]
Yep, that would have essentially the same result since PyGlossary's logic is based on dictutil. #14 is quite high up on my to-do list, but I haven't had enough contiguous free time to work on it yet. I'll probably end up doing it towards the middle of this year.

**Comment by kylefeng28** [comment]
I've been playing around with generating Japanese/Chinese dictionaries for Kobo, though I didn't use dictutil for it (but the docs were very helpful). I found this repo which converts JMdict to a Kobo dictionary and it works quite well for me (Libra Colour, 4.45.23697): https://github.com/cessen/kobo_jp_dict

I believe the prefix algorithm for Japanese is basically:
- take the first two characters, trim spaces, convert hiragana to katakana, and check the first character. 
  - if the first character is kana, return as is (similar to Cyrllic handling)
    - e.g. `あ` -> `あ` or`ア`, `あの` -> `あの` or `アノ`, ` あの` -> `あ` or `ア`
  - if the first character is kanji, return only that character
    - e.g. `日本語` -> `日`, `人々` -> `人`, `見る` -> `見`

I submitted a change for pyglossary that implements this, maybe dictutil can also have something similar?: https://github.com/ilius/pyglossary/pull/740

I also submitted a change for dictword-test with a `-j` flag that calls `DictionaryParser::htmlForWord(word, true)` where `true` represents using the Japanese lookup logic. I haven't rigorously reverse engineered it yet but the rules I mentioned above seem to hold true: https://github.com/pgaskin/kobo-mods/pull/8

(Edit, seems like the katakana conversion is needed after all, reverting my previous edits)

**Comment by ilius** [comment]
> convert hiragana to katakana

WOW, is this really needed?!

**Comment by BoboTiG** [comment]
My two cents, since I completely forgot that I commented on this issue.

I had quite a success with Japanese on Kobo (you can try https://www.reader-dict.com/download/ja if you want).

First above all, I saw that variants won't work, so duplication might be needed.
About the process, you can have a look at the code directly: https://github.com/reader-dict/monolingual/blob/b61b5480728cdbbb438e1c9aac3720df4a2048ac/wikidict/utils.py#L342

I am not sure we should do something at the PyGlossary level.

All in all, there might be more work to do on my side, but since I do not read Japanese, I'm waiting on readers feedback and I had only good news so far, I guess it is working fine then :)

**Comment by kylefeng28** [comment]
> WOW, is this really needed?!

Actually I don't think this is strictly needed for generating the filenames. (The PR I submitted for pyglossary didn't have this and it works fine). The dictionary lookup logic for Japanese seems to do some smart matching and handles both hiragana/katakana equivalents of the same word; maybe this katakana normalization is similar to how we normalize Latin words to lowercase to handle both uppercase/lowercase variants?

e.g. I checked the builtin `dicthtml-jaxxdjs.zip` file and it has both `あい.html` (hiragana) and a `アイ.html` (katakana). Both are pronounced "ai" so they should be considered equivalent.

Both files have basically the same duplicated kanji entries (愛/合い/藍/etc, all pronounced "ai"). However, the katakana version `アイ.html`  has some more entries since katakana is also used for foreign loanwords. Thus `アイ.html` has the English loanwords "eye" and "I"  whereas `あい.html` does not.

~Searching text for `あい` or `アイ` seem to bring up the same results, including "eye" and "I", so it seems like Kobo looks at both hiragana and katakana and merges the result.~

~So from a filename perspective, it seems we can generate either `あい` or `アイ` and the lookup logic will handle it~

(Edit: I was completely wrong about this, see my comment below)

**Comment by kylefeng28** [comment]
@BoboTiG thanks for linking the reader.dict website. I just downloaded the Japanese monolingual dictionary (`dicthtml-ja-ja.zip`) but I found some issues, even for words that are not variants. The kanji search seems to work as expected, but if the search term only has hiragana/katakana, it seems to break:

- Searched for hiragana `あい` (ai) -> showing results only for katakana `アイ`, even though there is a `あい.html` file with expected entries
- Searched for `サボる` (saboru, mixed hiragana-katakana) -> only shows entry `サボ` even though there is an entry for `サボる` in `サボ.html`
- Searched for `あらわれる` -> no results, even though there is a `あら.html` file with the expected entry

The kobo_jp_dict ja_en dictionaries work fine for hiragana/katakana search, and the reader.dict html files have the expected structure `<a name="... />` for entries, so I'm not sure what's going on. Perhaps there's some parsing issue going on with nested `<p>` tags?

Let me know if I should open in issue in reader.dict monolingual repo to track this issue, happy to investigate further

**Comment by pgaskin** [comment]
> First above all, I saw that variants won't work, so duplication might be needed.

Have you tried the prefix exceptions thing I mentioned in #14 for the updated format?

**Comment by BoboTiG** [comment]
> Let me know if I should open in issue in reader.dict monolingual repo to track this issue, happy to investigate further

Thank you for your tests. Yes, you could open an issue, and if you have hints to tackle the problem, I would love to know them :)

**Comment by BoboTiG** [comment]
> > First above all, I saw that variants won't work, so duplication might be needed.
> 
> Have you tried the prefix exceptions thing I mentioned in [#14](https://github.com/pgaskin/dictutil/issues/14) for the updated format?

I will take a look, it seems interesting and I also completely forgot about that. If that's a list of word redirection, that would be quite some improvement for us.

**Comment by kylefeng28** [comment]
>>> convert hiragana to katakana
>
>> WOW, is this really needed?! 
>
> Actually I don't think this is strictly needed for generating the filenames
> e.g. I checked the builtin `dicthtml-jaxxdjs.zip` file and it has both `あい.html` (hiragana) and a `アイ.html` (katakana)
> Searching text for `あい` or `アイ` seem to bring up the same results, including "eye" and "I", so it seems like Kobo looks at both hiragana and katakana and merges the result.

Update: After a lot more testing, it looks like I was completely wrong. The hiragana to katakana conversion is needed after all; Kobo only looks at the katakana file (`アイ.html`) and ignores the hiragana one (`あい.html`).

In addition, the `<a name="..." />` also needs to be in katakana. So we would have `アイ.html` with `<w><p><a name="アイ" /><b>あい</b>...</p></w>`

**Comment by pgaskin** [comment]
Also see https://github.com/pgaskin/dictutil/issues/14#issuecomment-3418339169 @BoboTiG, since you now only support 4.24.15676 due to the v3 dict format (and the fw version with the proper prefix_exceptions handling since the first one which had it was buggy).

**Comment by BoboTiG** [comment]
I do not know what to do with the additional number in the trie. From my tests, it also works without this number.
Do you think it is a crucial thing? Or should we use `1` by default for all entries?
